File: 189/solution.go
package _189

func rotate(nums []int, k int) {
	var reverse func(l, h int)
	reverse = func(l, h int) {
		for l < h {
			nums[l], nums[h] = nums[h], nums[l]
			l++
			h--
		}
	}
	n := len(nums)
	k %= n
	reverse(0, n-1)
	reverse(0, k-1)
	reverse(k, n-1)
}

File: 189/Solution.java
class Solution {
    private int[] nums;

    private void reverse(int l, int h) {
        while (l < h) {
            int tmp = nums[l];
            nums[l] = nums[h];
            nums[h] = tmp;
            l++;
            h--;
        }
    }

    public void rotate(int[] nums, int k) {
        this.nums = nums;
        int n = nums.length;
        k %= n;
        reverse(0, n - 1);
        reverse(0, k - 1);
        reverse(k, n - 1);
    }
}